### PR TITLE
clarify retry logic and test it

### PR DESCRIPTION
**Pull request by lukiano**

### CODE REVIEW

File: src/core/task.ts [hunk 1]
```diff
@@ -102,7 +102,13 @@ export class Task<T> {
    * with the given backoff strategy.
    */
   retry(backoff: Backoff, maxRetries: number): Task<T> {
-    return this._retry(backoff, maxRetries, 0);
+    if (maxRetries < 0) {
+      throw new Error('assertion failed: number of task retries must be greater than or equals zero');
```
Comment (srijs) [inline src/core/task.ts]: Should this return a `Task.fail` instead of throwing?

Comment (lukiano) [inline src/core/task.ts]: I thought about that when writing it, but line of thought was that if you pass an invalid argument you would want it to fail asap.

Comment (srijs) [inline src/core/task.ts]: That makes sense. Unfortunately it still means that Task composition would no longer be side-effect free.

Maybe could we find a side-effect free strategy for it instead. Perhaps just don't retry if it's negative? Not sure..

Comment (lukiano) [inline src/core/task.ts]: I like that <0 doesn't retry either. I've added a doco for that behavior.

**Comment by srijs** [comment]
100% \o/

**Pull request merged**